rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/uri/FallbackResolver.java

```diff
@@ -250,5 +274,49 @@ public void unwatch(ISourceLocation root, Consumer<ISourceLocationChanged> watch
         getClient().removeWatcher(root, watcher, getServer());
 
     }
+    
+    public boolean isFileManaged(ISourceLocation file) {
+        for (final var service : textDocumentServices) {
+            if (service.isManagingFile(file)) {
+                return true;
+            }
+        }
+        return false;
+    }
 
+    @Override
+    public ISourceLocation resolve(ISourceLocation input) throws IOException {
+        if (isFileManaged(input)) {
+            try {
+                // The offset/length part of the source location is stripped off here.
+                // This is reinstated by `URIResolverRegistry::resolveAndFixOffsets`
+                // during logical resolution
+                return URIUtil.changeScheme(input, "lsp+" + input.getScheme()).top();
+            } catch (URISyntaxException e) {
+                // fall through
+            }
+        }
+        return input;
+    }
+
+    @Override
+    public String authority() {
+        throw new UnsupportedOperationException("'authority' not supported by fallback resolver");
+    }
+
+    private final List<IBaseTextDocumentService> textDocumentServices = new CopyOnWriteArrayList<>();
+
+    public void registerTextDocumentService(IBaseTextDocumentService service) {
+        textDocumentServices.add(service);
+    }
+
+    public TextDocumentState getDocumentState(ISourceLocation file) throws IOException {
+        for (final var service : textDocumentServices) {
+            final var state = service.getDocumentState(file);
```

Comment:
nit: why are they marked final in this file? (also in the other loop?)

Reply:
No particular reason. They've been removed for a more uniform code base.